src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1104 @@
+    if (mDelegate->IsProgressElement(aLocationId))
+    {
+        ChipLogError(Zcl, "AddPendingProgressElement - progress element already exists for location %u", aLocationId);
+        return false;
+    }
+
+    if (!mDelegate->AddProgressElement(inactiveProgress, dummyIndex))
+    {
+        return false;
+    }
+
+    NotifyProgressChanged();
+    return true;
+}
+
+bool Instance::SetProgressStatus(uint32_t aLocationId, OperationalStatusEnum opStatus)
+{
+    uint32_t listIndex;
+
+    Structs::ProgressStruct::Type progressElement;
+
+    if (!mDelegate->GetProgressElementById(aLocationId, listIndex, progressElement))
+    {
+        ChipLogError(Zcl, "SetProgressStatus - progress element does not exist for location %u", aLocationId);
+        return false;
+    }
+
+    // If the status value is not changing, there in no need to modify the existing element.
+    if (progressElement.status == opStatus)
+    {
+        return true;
+    }
+
+    // set the progress status in the local copy
+    progressElement.status = opStatus;
+
+    // TotalOperationalTime SHALL be null if the Status field is not set to Completed or Skipped.
+    if ((opStatus != OperationalStatusEnum::kCompleted) && (opStatus != OperationalStatusEnum::kSkipped))
+    {
+        progressElement.totalOperationalTime.Value().SetNull();
+    }
+
+    // add the updated element to the progress attribute
+    if (!mDelegate->ModifyProgressElement(listIndex, progressElement))
+    {
+        return false;
+    }
+
+    NotifyProgressChanged();
+    return true;
+}
+
+bool Instance::SetProgressTotalOperationalTime(uint32_t aLocationId, const DataModel::Nullable<uint32_t> & aTotalOperationalTime)
+{
+    uint32_t listIndex;
+
+    Structs::ProgressStruct::Type progressElement;
+
+    if (!mDelegate->GetProgressElementById(aLocationId, listIndex, progressElement))
+    {
+        ChipLogError(Zcl, "SetProgressTotalOperationalTime - progress element does not exist for location %u", aLocationId);
+        return false;
+    }
+
+    // If the time value is not changing, there is no need to modify the existing element.
+    if (progressElement.totalOperationalTime == aTotalOperationalTime)
+    {
+        return true;
+    }
+
+    // This attribute SHALL be null if the Status field is not set to Completed or Skipped
+    if (((progressElement.status == OperationalStatusEnum::kCompleted) ||
+         (progressElement.status == OperationalStatusEnum::kSkipped)) &&
+        !aTotalOperationalTime.IsNull())
+    {
+        ChipLogError(Zcl,
+                     "SetProgressTotalOperationalTime - location %u opStatus value %u - can be non-null only if opStatus is "
+                     "Completed or Skipped",
+                     aLocationId, to_underlying(progressElement.status));
+        return false;
+    }
+
+    // set the time in the local copy
+    progressElement.totalOperationalTime.Value() = aTotalOperationalTime;
```

Comment:
Wait, what if `progressElement.totalOperationalTime` doesn't have `Value()` right now?  This will crash, no?

This should probably `Emplace` instead.

Reply:
fixed [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/e68858d16fb395129225cbaa1cd8c51f787b39be).